pkg/cleaner/azure/resourcegroup.go

```diff
@@ -0,0 +1,104 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-02-01/resources"
+	"github.com/giantswarm/microerror"
+)
+
+const (
+	// gracePeriod represents the maximum time the CI resources are allowed to
+	// remain up. CI resources older than gracePeriod will be deleted.
+	gracePeriod = 90 * time.Minute
+)
+
+func (c Cleaner) cleanResourceGroup(ctx context.Context) error {
+	// It would be more efficient here to use a filter like "startswith(name,'ci-') or startswith(name,'e2e')"
+	// but this does not seems to work now, see https://github.com/Azure/azure-sdk-for-go/issues/2480.
+	groupIter, err := c.groupsClient.ListComplete(ctx, "", nil)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deadLine := time.Now().Add(-gracePeriod).UTC()
+
+	for ; groupIter.NotDone(); groupIter.Next() {
+		group := groupIter.Value()
+
+		c.logger.Log("level", "debug", "message", fmt.Sprintf("checking resource group %q", *group.Name))
+
+		shouldBeDeleted, err := c.groupShouldBeDeleted(ctx, group, deadLine)
+		if err != nil {
+			c.logger.Log("level", "debug", "message", fmt.Sprintf("skipping resource group %q due to error", *group.Name), "error", err.Error())
+			continue
+		}
+
+		if shouldBeDeleted {
+			respFuture, err := c.groupsClient.Delete(ctx, *group.Name)
+			if err != nil {
+				c.logger.Log("level", "error", "message", fmt.Sprintf("resource group %q deletion failed", *group.Name), "error", err.Error())
+				return microerror.Mask(err)
```

Comment:
We could not return here, just save the error and continue to the next item. Then at the end return the error if any happened, WDYT? If not and this concrete item deletion continues to fail, the rest of items in the list won't be deleted.

Reply:
Good point. We should always continue and try to clean as much as possible.

We are then making a conscious decision of discard every error which happened but the last, fine with me as there is no way to handle multiple errors here.

Reply:
That's up to you, you can have a slice of errors and handle them as you wish, return the first, return the last, return a new one with messages from all of them...

Reply:
updated this, PTAL